rust/kernel/platform_driver.rs

```diff
@@ -0,0 +1,167 @@
+// SPDX-License-Identifier: GPL-2.0
+
+use crate::error::{Error, KernelResult};
+use crate::{bindings, c_types, CStr};
+use alloc::boxed::Box;
+use core::marker::PhantomPinned;
+use core::mem::transmute;
+use core::pin::Pin;
+
+extern "C" {
+    #[allow(improper_ctypes)]
+    fn rust_helper_platform_get_drvdata(
+        pdev: *const bindings::platform_device,
+    ) -> *mut c_types::c_void;
+
+    #[allow(improper_ctypes)]
+    fn rust_helper_platform_set_drvdata(
+        pdev: *mut bindings::platform_device,
+        data: *mut c_types::c_void,
+    );
+}
+
+unsafe extern "C" fn probe_callback<T: PlatformDriver>(
+    pdev: *mut bindings::platform_device,
+) -> c_types::c_int {
+    let f = || {
+        let drv_data = T::probe(&mut PlatformDevice::new(pdev))?;
+        let drv_data = Box::try_new(drv_data)?;
+        let drv_data = Box::into_raw(drv_data) as *mut c_types::c_void;
+        Ok(drv_data) as KernelResult<_>
+    };
+    // TODO don't we need to pin this?
+    let ptr = match f() {
+        Ok(ptr) => ptr,
+        Err(e) => return e.to_kernel_errno(),
+    };
+    rust_helper_platform_set_drvdata(pdev, ptr);
+    0
+}
+
+fn new_of_device_id(compatible: &CStr<'static>) -> KernelResult<bindings::of_device_id> {
+    // TODO:
+    // - fail at build time if compatible CStr doesn't fit.
+    // - can we do this safely without transmute?
+    let mut buf = [0_u8; 128];
+    if compatible.len() > buf.len() {
+        return Err(Error::EINVAL);
+    }
+    // PANIC: this will never panic: `compatible` is not longer than `buf`.
+    buf[..compatible.len()].copy_from_slice(compatible.as_bytes());
+    Ok(bindings::of_device_id {
+        // SAFETY: re-interpretation from [u8] to [i8] of same length is always safe.
+        compatible: unsafe { transmute::<[u8; 128], [i8; 128]>(buf) },
+        ..Default::default()
+    })
+}
+
+unsafe extern "C" fn remove_callback<T: PlatformDriver>(
+    pdev: *mut bindings::platform_device,
+) -> c_types::c_int {
+    let ptr = rust_helper_platform_get_drvdata(pdev);
+    let drv_data: Box<T::DrvData> = Box::from_raw(ptr as _);
+    drop(drv_data);
+    0
+}
+
+/// A registration of a platform driver.
+#[derive(Default)]
+pub struct Registration {
+    registered: bool,
+    pdrv: bindings::platform_driver,
+    of_table: [bindings::of_device_id; 2],
+    _pin: PhantomPinned,
+}
+
+impl Registration {
+    fn register<P: PlatformDriver>(
+        self: Pin<&mut Self>,
+        name: CStr<'static>,
+        of_id: CStr<'static>,
+        module: &'static crate::ThisModule,
+    ) -> KernelResult {
+        // SAFETY: We must ensure that we never move out of `this`.
+        let this = unsafe { self.get_unchecked_mut() };
+        if this.registered {
+            // Already registered.
+            return Err(Error::EINVAL);
+        }
+        // TODO should create a variable size table here.
+        this.of_table[0] = new_of_device_id(&of_id)?;
```

Comment:
Take a look at how we do this for chrdev: https://github.com/Rust-for-Linux/linux/blob/rust/rust/kernel/chrdev.rs#L23

We use const generics to size the registration, and then you can call `register()` up to `N` times

Reply:
Thanks !! I did try something along those lines. I ran into the following issue: an `of` table with `N` compatible strings has `N+1` entries. When I tried to use `N+1`, rustc told me to activate some extra features. And when I switched those on, rustc told me it could result in compiler crashes...

Reply:
Some of the `const_fn` features are still unstable, but I think they solve this well, so I think it makes sense to attempt to try that out -- if it's geniunely non-functional, then we could try a different approach.

Reply:
Thanks ! I believe at least two alternatives were suggested during last Saturday's call: `const_fn` and procedural macros. Does it make sense to prefer one over the other, i.e. which one to try first?

@Kloenk expressed interest in writing a procmac for `of_table`. Finn, are you still ok with that? Would you like to try out @alex's `const_fn` suggestion as well?

There's probably more here than meets the eye. `of_table`'s `data` member allows drivers to distinguish between device "flavours" and adjust accordingly. For example:
```c
static const struct of_device_id bcm2835_rng_of_match[] = {
	{ .compatible = "brcm,bcm2835-rng"},
	{ .compatible = "brcm,bcm-nsp-rng", .data = &nsp_rng_of_data },
	{ .compatible = "brcm,bcm5301x-rng", .data = &nsp_rng_of_data },
	{ .compatible = "brcm,bcm6368-rng"},
	{},
};
```
The `data` above indicates that the `5301` and `nsp-rng` variants require interrupt masking. I guess the appropriate Rust abstraction here would be an `enum`? Pass an optional enum variant for every table entry?

Reply:
> Does it make sense to prefer one over the other, i.e. which one to try first?

In general, it is best to avoid macros, in particular proc ones. We should reach for them when there is no other way to do something.

We may also want to use them if they really improve/simplify some interface for users or if they allow for a much easier/cleaner implementation. But, in general, if something can be done without them, it is usually the best way.

Reply:
@Kloenk let me know if you would like to look into a `const_fn` way of doing this.